### ISSUE TITLE
No Tiff support.

### ISSUE THREAD

**Issue by JimBobSquarePants** [SixLabors/ImageSharp#12]
_From @JimBobSquarePants on August 31, 2016 1:49_

We should be able to support TIFF

Useful links.

http://partners.adobe.com/public/developer/tiff/index.html

http://www.remotesensing.org/libtiff/

http://www.awaresystems.be/imaging/tiff.html

Current implementation can be found in the [tiff-codec](https://github.com/SixLabors/ImageSharp/tree/tiff-codec) branch.

### Feature Checklist

- [x] Reading TIFF file structure
- [x] Parsing of relevant metadata
- [x] Extraction of image data blocks
- [x] Construction of image (from strip data)
- [ ] Construction of image (from tile data)

### Compression Types

- [x] None
- [ ] Ccitt1D
- [x] PackBits
- [ ] CcittGroup3Fax
- [ ] CcittGroup4Fax
- [x] Lzw
- [ ] Old Jpeg
- [ ] Jpeg (Technote 2)
- [x] Deflate (Technote 2)
- [x] Old Deflate (Technote 2)

### Photometric Interpretation Formats

- [x] WhiteIsZero
- [x] BlackIsZero
- [x] Rgb (Chunky)
- [x] Rgb (Planar)
- [x] PaletteColor
- [ ] TransparencyMask
- [ ] Separated (TIFF Extension color spaces)
- [ ] YCbCr (TIFF Extension color spaces)
- [ ] CieLab (TIFF Extension color spaces)
- [ ] IccLab (TechNote 1)

**Comment by JimBobSquarePants** [comment]
_From @kierenj on September 7, 2016 20:27_

I realise how unhelpful this comment right here is, but TIFF decoding across the board can be really tricky to get right - there are loads of compression formats, and even fairly respected tools (from MS / Atalasoft) get it wrong even with 1-bit images.  I think some archiving formats even require special licensing to implement correctly.  It might be worth identifying a small subset of formats to support!

**Comment by JimBobSquarePants** [comment]
@kierenj It's certainly not an unhelpful comment. Tiff would be hard and I'm sorely tempted to let someone else figure it out and concentrate on what we have already. I'm just dreading the day someone tells me that the library is useless without it. There's always someone!

**Comment by chrisvandijk** [comment]
I needed to import Tiff files so I created a temporary [Tiff encoder and decoder](https://github.com/chrisvandijk/ImageSharp-Tiff) based on LibTiff.Net.

**Comment by JimBobSquarePants** [comment]
@chrisvandijk Good effort but have you benchmarked that codec? The Bitmiracle jpeg codec was about 10x slower and used 190MB more to process a small test image compared to our codec. I imagine the Tiff one will be the same. I'd be careful using it. 

@Andy-Wilkinson you have been playing around with Tiff haven't you with your own stuff? I wonder if we can use some of it?

**Comment by chrisvandijk** [comment]
@JimBobSquarePants The code and performance is bad, but I really needed the ability to open Tiff files. It's a temporary solution until a better one.

**Comment by JimBobSquarePants** [comment]
@chrisvandijk understandable. Tiff is a complex format so I'm pushing it to the backburner until I can get everything else out of the way... Not looking forward to tackling it!

**Comment by antonfirsov** [comment]
@chrisvandijk @JimBobSquarepants One big "advantage" of these naive C ports is that its really not that hard to make the first steps towards speeding & cleaning them up. 
Someone with time & a profiler could check it, and estimate the efforts needed reuse/refactor elements of that code.

**Comment by Andy-Wilkinson** [comment]
@JimBobSquarePants I've got the basics of reading the Tiff format itself working from my own project (I started here since many Camera Raw files are mutant Tiff files). As you mention, there are a number of compression formats, but the key ones probably aren't too much trouble. The one issue with my existing codebase is that it was a bit of an experiment in a more functional style of C#, rather than how you might structure it normally.

All that said, I know a fair amount about the Tiff format now, so probably would be able to make a quick start on a true ImageSharp Tiff codec pretty easily (and share a bit of the code where I can). Would you be interested in me picking up this issue and giving it a start? (especially since you guys seem to be busy enough with the existing codec perf 😄)

**Comment by JimBobSquarePants** [comment]
@antonfirsov I've found with those ports though there's about a million lines more code than I'd like to read nested in there.

@Andy-Wilkinson That would be absolutely brilliant if you could! 💯

**Comment by sandorfr** [comment]
I come across this library https://github.com/BitMiracle/libtiff.net which is already .netstandard compatible. 
It's under new BSD license https://github.com/BitMiracle/libtiff.net/blob/master/license.txt

I feel like this is a good candidate to bring support for TIFF. 

What do you think?

**Comment by JimBobSquarePants** [comment]
@sandorfr We've already got a WIP TIFF PR in progress #119 

The BitMiracle repositories, to put it nicely, are junk. Poorly ported, slow, bloated implementations that are useful only for testing against.

**Comment by poeman16** [comment]
Any updates on https://github.com/SixLabors/ImageSharp/pull/119?  Thanks!

**Comment by Andy-Wilkinson** [comment]
@poeman16 : Unfortunately I've had pretty much zero time to devote to working on this pull request for a while. The foundations of parsing the TIFF structure was mostly in place, with the most complex area remaining being implementations for the different compression formations, in particular those based on JPEG. There would need to be some changes to fit with any changes in the ImageSharp design since the last commit. Definitely something that could be picked up by someone.

**Comment by JimBobSquarePants** [comment]
There's been a lot of changes in the main repo since the PR was opened. We'll have to do a lot of work to update the codebase before support can be completed.

**Comment by darl0026** [comment]
Are there any plans to support TIFF in the future, im using your library for viewing/editing Exif data

**Comment by damianog** [comment]
Hi @JimBobSquarePants 
as the others I am interested on a Tiff decoder. I need to process multipage Tiff documents from office scanner.
Thanks in advance.

**Comment by IldarKhayrutdinov** [comment]
@antonfirsov @JimBobSquarePants @Andy-Wilkinson  Can I contribute to the _tiff-codec_ branch?

**Comment by antonfirsov** [comment]
@IldarKhayrutdinov that would be ace! 🚀

**Comment by claunia** [comment]
Just a comment, I don't see it named, but there are multipage TIFF files, used extensively by a lot of scanning software, so should be accounted for?

**Comment by IldarKhayrutdinov** [comment]
@claunia Yes, at the moment the multipages aren't supported, I'm working on it ☺

**Comment by Andy-Wilkinson** [comment]
Thanks @IldarKhayrutdinov for picking up the TIFF decoder. I have to confess that I've not had much input into this for a while, but I think @JimBobSquarePants has made a few updates to keep things moving along.

The biggest piece of work from what I remember was implementing all the more common compression types (see the checklist at the start of #12). Multipage TIFFs was something that was going to be a bit tricky with the rest of ImageSharp at the time, but this may be different now. There's probably a lot of performance improvements that could be done too!

Good luck... :-)

**Comment by IldarKhayrutdinov** [comment]
Thanks @Andy-Wilkinson You did a very good job! I'd be honored  to continue your contribution slightly_smiling_face

**Comment by erikarenhill** [comment]
Any updates to this issue, what's left to do, in need of more help?

**Comment by JimBobSquarePants** [comment]
@IldarKhayrutdinov Over to you... 😄

**Comment by IldarKhayrutdinov** [comment]
@erikarenhill @JimBobSquarePants At the moment, I have local changes and continue working, not fast, but I continue :slightly_smiling_face: If you wish, you can add support for example Jpeg (Technote 2), I think, our code will not be difficult to merge into one

**Comment by brianpopow** [comment]
I have added support for decompressing `Ccitt1D` and `CcittGroup3Fax`. 

There is still an issue with `Ccitt1D` for large images, though, but i will find that bug.
See the failing test: `Tiff/Calliphora_huffman_rle.tif`

**Comment by brianpopow** [comment]
I have worked on the tiff encoder the last few days. The following works so far:

- [x] Encode uncompressed RGB
- [x] Encode uncompressed gray
- [x] Encode paletted color images

I will continue with the different compression methods.

**Comment by ptasev** [comment]
Is there support for reading or writing custom tiff tags (I'm assuming they're the same as exif tags?).  I was looking at the metadata code, but I see methods like GetString or GetArray are internal, and I didn't see a way to add new tags and data either.

**Comment by brianpopow** [comment]
> Is there support for reading or writing custom tiff tags (I'm assuming they're the same as exif tags?). I was looking at the metadata code, but I see methods like GetString or GetArray are internal, and I didn't see a way to add new tags and data either.

We keep it the same as with EXIF, no custom tags are allowed. Only tags which are in the spec.

**Comment by ptasev** [comment]
That's a shame. It is quite useful to add extra data to images (per frame). It's helpful to allow us to label images in the biotech industry.

**Comment by antonfirsov** [comment]
> It is quite useful to add extra data to images (per frame). It's helpful to allow us to label images in the biotech industry.

@JimBobSquarePants @dlemstra (Without knowing anything about the image metadata domain) I wonder how common this practice is, and how much does it go against standards? Maybe we should consider it, if it blocks (some) users?

**Comment by IldarKhayrutdinov** [comment]
@antonfirsov I came across files with custom tiff tags. 
The spec seems to allow it, quote from [TIFF6.pdf](https://www.adobe.io/content/dam/udp/en/open/standards/tiff/TIFF6.pdf) (page 8):
> However, if there is little or no chance that your TIFF files
> will escape your private environment, please consider using TIFF tags in the
> “reusable” 65000-65535 range.

**Comment by brianpopow** [comment]
@IldarKhayrutdinov Hi Ildar. As far as i know, you did most of the LZW implementation. While it works for some image, it has some issues with other images for example `rgb_lzw_no_predictor.tiff` (its in the Tiff test file directory, i cannot attach it here). Here are some info's about the image:

```
tiffinfo rgb_lzw_no_predictor.tiff
TIFF Directory at offset 0x1fed4 (130772)
  Image Width: 256 Image Length: 256
  Resolution: 300, 300 pixels/inch
  Bits/Sample: 8
  Compression Scheme: LZW
  Photometric Interpretation: RGB color
  FillOrder: msb-to-lsb
  Orientation: row 0 top, col 0 lhs
  Samples/Pixel: 3
  Rows/Strip: 256
  Planar Configuration: single image plane
  Page Number: 0-1
  White Point: 0.3127-0.329
  PrimaryChromaticities: 0.640000,0.330000,0.300000,0.600000,0.150000,0.060000
  Predictor: none 1 (0x1)
``` 

The decoded image is mostly black. I have trouble finding the issue. It would be great if you could review this once again and help me finding this. All i can say so far, is that the while loop breaks after only 428 pixels are decoded.

**Comment by IldarKhayrutdinov** [comment]
@brianpopow Hi Brian! You've implemented so much for TIFF 😃 It seems `rgb_lzw_no_predictor.tiff` file is corrupted.

**Comment by brianpopow** [comment]
> @brianpopow Hi Brian! You've implemented so much for TIFF 😃 It seems `rgb_lzw_no_predictor.tiff` file is corrupted.

@IldarKhayrutdinov Thanks alot for looking into this. It seems, i did mess up the image when pushing it to git :confused:

I have attached it here as a zip file:
[rgb_lzw_no_predictor.zip](https://github.com/SixLabors/ImageSharp/files/5628593/rgb_lzw_no_predictor.zip)

Im sorry for the inconvenience. I hope you could look into this again. I think you can spot an error faster here with LZW then i can.

**Comment by IldarKhayrutdinov** [comment]
@brianpopow yeah, now the bug is being reproduced, I will investigate it today

**Comment by JimBobSquarePants** [comment]
>@brianpopow Hi Brian! You've implemented so much for TIFF 😃 It seems rgb_lzw_no_predictor.tiff file is corrupted.

I think our `.gitattributes` might be breaking tiff files. We should add the extensions for all known image formats.

**Comment by JimBobSquarePants** [comment]
Just pushed changes to fix that. Once #1451 is merged you can update from master

**Comment by IldarKhayrutdinov** [comment]
> @brianpopow yeah, now the bug is being reproduced, I will investigate it today

@brianpopow done! #1457

**Comment by brianpopow** [comment]
@IldarKhayrutdinov amazing, great job! I will check it out and test it.

**Comment by brianpopow** [comment]
I am happy to announce that we have now support for decoding and encoding Tiff images with #1553 merged to the master branch.

You can try it out with our nightly build from our [myget feed](https://www.myget.org/feed/sixlabors/package/nuget/SixLabors.ImageSharp)

For now this feature has to be activated manually by the user, which can be done in several different ways:

1. Add Tiff to the Default Configuration:

```
Configuration.Default.AddTiff();
```

2. Or to your specific configuration:

```
Configuration config = new Configuration();
config.AddTiff();
```

3. Provide the Tiff Decoder / Encoder explicit when loading or saving a image:

```
using var image = Image.Load(@"image.tiff", new TiffDecoder());
```

**Comment by JimBobSquarePants** [comment]
@brianpopow let’s make that automatic. 🙂

**Comment by brianpopow** [comment]
> @brianpopow let’s make that automatic. 🙂

With #1634 merged, the Tiff format is now added to the default Configuration.

**Comment by cocosip** [comment]
Hi, i try to read a tiff image, but it throw `Tiled images are not supported`. 
Any plan to support `Construction of image (from tile data)` ?
Thank you!

**Comment by JimBobSquarePants** [comment]
No plans currently. That would require community support to implement.
